### ISSUE TITLE
ElixirLS Dialyzer (VS code) warning : The pattern 'false' can never match the type  'true'

### ISSUE THREAD

**Issue by relax2code** [Qqwy/elixir-type_check#85]
Hi,  first I want to thank you for this great library , I have some issue with Dialyzer warning but I'm not sure why , here's my settings:

Mix.exs: 
```elixir
      {:type_check, github: "Qqwy/elixir-type_check"}
```

Elixir vers:
   1.12.2-otp-24

ElixirLS vs code: 
  0.8.1

Code:

```elixir
defmodule Document.ExcelTest do
  use TypeCheck
  alias Document.ExcelTest

  defstruct [
    :name,
    :age,
    :phone
  ]

  @type! t() :: %ExcelTest{
           name: String.t(),
           age: integer(),
           phone: String.t()
         }
end
```
```elixir
defmodule Document.Doc do

  use TypeCheck
  alias Document.ExcelTest

  @spec! parse_list(any) :: ExcelTest.t()
  def parse_list(list) do

    column = [
      :name,
      :age,
      :phone
    ]

    Enum.zip(column, list)
    |> Enum.into(%{})
    |> then( fn map -> struct(%ExcelTest{}, map) end)

  end
end
```
 the code can be compiled and run  , but there' warning from dialyzer ;
 
![screenshot](https://user-images.githubusercontent.com/5886490/143420717-7fd808bd-e41f-4db1-abf1-97d097745344.png)

Is there something wrong with my code ?

**Comment by Qqwy** [comment]
:thinking: Hmm, I wonder whether Dialyzer is complaining about some code inside the runtime typecheck that is generated.

Thanks for filing this bugreport, I'll look into it!

**Comment by tomekowal** [comment]
I have a similar issue. The code:

```elixir
defmodule ExKeymap.KeymapItem do
  use TypeCheck
  defstruct [:item_name, :help, :fun]

  @type! t :: %__MODULE__{
    item_name: String.t(),
    fun: (-> any()),
    help: String.t() | nil
  }

  @spec! new(String.t(), (-> any()), String.t() | nil) :: t()
  def new(item_name, fun, help \\ nil) do
    %__MODULE__{
      item_name: item_name,
      fun: fun,
      help: help
    }
  end
end
```

The output:

```
mix dialyzer
Finding suitable PLTs
Checking PLT...
[:compiler, :elixir, :iex, :kernel, :logger, :stdlib, :type_check]
PLT is up to date!
No :ignore_warnings opt specified in mix.exs and default does not exist.

Starting Dialyzer
[
  check_plt: false,
  init_plt: '/home/tomaszkowal/experiments/ex_keymap/_build/dev/dialyxir_erlang-24.1.2_elixir-1.12.3_deps-dev.plt',
  files: ['/home/tomaszkowal/experiments/ex_keymap/_build/dev/lib/ex_keymap/ebin/Elixir.ExKeymap.Keymap.beam',
   '/home/tomaszkowal/experiments/ex_keymap/_build/dev/lib/ex_keymap/ebin/Elixir.ExKeymap.KeymapItem.beam',
   '/home/tomaszkowal/experiments/ex_keymap/_build/dev/lib/ex_keymap/ebin/Elixir.ExKeymap.beam',
   '/home/tomaszkowal/experiments/ex_keymap/_build/dev/lib/ex_keymap/ebin/Elixir.TypeCheck.Internals.UserTypes.ExKeymap.Keymap.beam',
   '/home/tomaszkowal/experiments/ex_keymap/_build/dev/lib/ex_keymap/ebin/Elixir.TypeCheck.Internals.UserTypes.ExKeymap.KeymapItem.beam'],
  warnings: [:unknown]
]
Total errors: 1, Skipped: 0, Unnecessary Skips: 0
done in 0m0.68s
lib/type_check/spec.ex:1:pattern_match
The pattern can never match the type.

Pattern:
false

Type:
true

________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```

**Comment by Qqwy** [comment]
An even more minimal version that results in the Dialyzer warning:

```elixir
defmodule ExKeymap.KeymapItem do
  use TypeCheck
  defstruct [:help]

  @type! t() :: %__MODULE__{
    help: binary() | nil
  }

  @spec! new(binary() | nil) :: t()
  def new(help) do
    %__MODULE__{
      help: help
    }
  end
end
```
